gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/task/DagProcessingEngineMetrics.java

```diff
@@ -0,0 +1,177 @@
+    updateMetricForDagActionType(this.dagActionsObservedMeterByDagActionType, dagActionType);
+  }
+
+  public void markDagActionsLeasedObtained(DagActionStore.LeaseParams leaseParams) {
+    updateMetricForDagActionType(this.dagActionsLeasesObtainedMeterByDagActionType,
+        leaseParams.getDagAction().getDagActionType());
+  }
+
+  public void markDagActionsNoLongerLeasing(DagActionStore.LeaseParams leaseParams) {
+    updateMetricForDagActionType(this.dagActionsNoLongerLeasingMeterByDagActionType,
+        leaseParams.getDagAction().getDagActionType());
+  }
+
+  public void markDagActionsLeaseReminderScheduled(DagActionStore.LeaseParams leaseParams) {
+    updateMetricForDagActionType(this.dagActionsLeaseReminderScheduledMeterByDagActionType,
+        leaseParams.getDagAction().getDagActionType());
+  }
+
+  public void markDagActionsRemindersProcessed(DagActionStore.LeaseParams leaseParams) {
+    updateMetricForDagActionType(this.dagActionsReminderProcessedMeterByDagActionType,
+        leaseParams.getDagAction().getDagActionType());
+  }
+
+  // TODO: implement evaluating max retries later
+  public void markDagActionsExceedingMaxRetry(DagActionStore.DagActionType dagActionType) {
+    updateMetricForDagActionType(this.dagActionsExceededMaxRetryMeterByDagActionType, dagActionType);
+  }
+
+  public void markDagActionsInitialize(DagActionStore.DagActionType dagActionType, boolean succeeded) {
+    if (succeeded) {
+      updateMetricForDagActionType(this.dagActionsInitializeSucceededMeterByDagActionType, dagActionType);
+    } else {
+      updateMetricForDagActionType(this.dagActionsInitializeFailedMeterByDagActionType, dagActionType);
+    }
+  }
+
+  public void markDagActionsAct(DagActionStore.DagActionType dagActionType, boolean succeeded) {
+    if (succeeded) {
+      updateMetricForDagActionType(this.dagActionsActSucceededMeterByDagActionType, dagActionType);
+    } else {
+      updateMetricForDagActionType(this.dagActionsActFailedMeterByDagActionType, dagActionType);
+    }
+  }
+
+  public void markDagActionsConclude(DagActionStore.DagActionType dagActionType, boolean succeeded) {
+    if (succeeded) {
+      updateMetricForDagActionType(this.dagActionsConcludeSucceededMeterByDagActionType, dagActionType);
+    } else {
+      updateMetricForDagActionType(this.dagActionsConcludeFailedMeterByDagActionType, dagActionType);
+    }
+  }
+
+
+  /**
+   * Generic helper used to increment a metric corresponding to the dagActionType in the provided map. It assumes the
+   * meter for each dagActionType can be identified by its name.
+   */
+  private void updateMetricForDagActionType(ConcurrentMap<String, ContextAwareMeter> metricMap,
+      DagActionStore.DagActionType dagActionType) {
+      if (metricMap.containsKey(dagActionType.toString())) {
+        metricMap.get(dagActionType.toString()).mark();
+      } else {
+        throw new RuntimeException(String.format("No meter exists for dagActionType %s in metricsMap %s",
```

Comment:
or swallow the exception?

Reply:
Kip argued that this should practically never happen so we should throw exception loudly if it's missing.

Reply:
in that case, we should just replace `if ... else ` with  `metricMap.get(dagActionType).mark();`